certifier/src/certifier.rs

```diff
@@ -20,71 +23,240 @@ pub struct CertifyRequest {
 }
 
 #[serde_as]
-#[derive(Debug, Serialize)]
-struct CertifyResponse {
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CertifyResponse {
+    /// The certificate as scale-encoded `Certificate` struct
     #[serde_as(as = "Base64")]
-    signature: Vec<u8>,
+    pub certificate: Vec<u8>,
+    /// Signature of the certificate
     #[serde_as(as = "Base64")]
-    pub_key: Vec<u8>,
+    pub signature: Vec<u8>,
+    /// The public key of the certifier that signed the certificate
+    #[serde_as(as = "Base64")]
+    pub pub_key: Vec<u8>,
+}
+
+#[derive(Debug, Decode, Encode)]
+pub struct Certificate {
+    // ID of the node being certified
+    pub pub_key: Vec<u8>,
+    /// Unix timestamp
+    pub expiration: Option<Compact<u64>>,
 }
 
 #[instrument(skip(state))]
 async fn certify(
-    State(state): State<Arc<AppState>>,
+    State(state): State<Arc<Certifier>>,
     Json(req): Json<CertifyRequest>,
 ) -> Result<Json<CertifyResponse>, (StatusCode, String)> {
     tracing::debug!("certifying");
 
-    let pub_key = req.metadata.node_id;
-    let my_id = state.signer.verifying_key().to_bytes();
     let s = state.clone();
+    let result = tokio::task::spawn_blocking(move || s.certify(&req.proof, &req.metadata))
+        .await
+        .map_err(|e| {
+            tracing::error!("internal error verifying proof: {e:?}");
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "error verifying proof".into(),
+            )
+        })?;
 
-    let result = tokio::task::spawn_blocking(move || {
-        s.verifier
-            .verify(&req.proof, &req.metadata, &s.cfg, &s.init_cfg, Mode::All)
-    })
-    .await
-    .map_err(|e| {
-        tracing::error!("internal error verifying proof: {e:?}");
-        (
-            StatusCode::INTERNAL_SERVER_ERROR,
-            "error verifying proof".into(),
-        )
-    })?;
-
-    result.map_err(|e| (StatusCode::FORBIDDEN, format!("invalid proof: {e:?}")))?;
-
-    // Sign the nodeID
-    let response = CertifyResponse {
-        signature: state.signer.sign(&pub_key).to_vec(),
-        pub_key: my_id.to_vec(),
-    };
-    Ok(Json(response))
+    match result {
+        Ok(result) => {
+            let response = CertifyResponse {
+                certificate: result.0.to_vec(),
+                signature: result.1.to_vec(),
+                pub_key: state.signer.verifying_key().to_bytes().to_vec(),
+            };
+            Ok(Json(response))
+        }
+        Err(e) => {
+            return Err((StatusCode::FORBIDDEN, format!("invalid proof: {e:?}")));
```

Comment:
I assume this should return an error like "not authorized" instead?

Reply:
Wdym? The `format!("invalid proof: {e:?}")` part is put in the response body and informs the caller why he was refused.